packages/composites/gm-token/src/transport/price.ts

```diff
@@ -0,0 +1,238 @@
+  }
+
+  async handleRequest(param: RequestParams) {
+    let response: AdapterResponse<BaseEndpointTypes['Response']>
+    try {
+      response = await this._handleRequest(param)
+    } catch (e) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(
+    param: RequestParams,
+  ): Promise<AdapterResponse<GmTokenTransportTypes['Response']>> {
+    const { index, long, short, market } = param
+
+    const assets = [index, long, short]
+
+    const providerDataRequestedUnixMs = Date.now()
+
+    const [indexPrices, longPrices, shortPrices] = await this.fetchPrices(
+      assets,
+      providerDataRequestedUnixMs,
+    )
+
+    const indexToken = tokenAddresses.arbitrum[index]
+    const longToken = tokenAddresses.arbitrum[long]
+    const shortToken = tokenAddresses.arbitrum[short]
+
+    const tokenPriceContractParams = [
+      this.settings.DATASTORE_CONTRACT_ADDRESS,
+      [market, indexToken, longToken, shortToken],
+      [indexPrices.ask, indexPrices.bid],
+      [longPrices.ask, longPrices.bid],
+      [shortPrices.ask, shortPrices.bid],
+      utils.keccak256(this.abiEncoder.encode(['string'], [this.settings.PNL_FACTOR_TYPE])),
+    ]
+
+    // Prices have a spread from min to max. The last param (maximize-true/false) decides whether to maximize the market token price
+    // or not. We get both values and return the median.
+    const [[maximizedValue], [minimizedValue]] = await Promise.all([
+      this.readerContract.getMarketTokenPrice(...tokenPriceContractParams, true),
+      this.readerContract.getMarketTokenPrice(...tokenPriceContractParams, false),
+    ])
+
+    const maximizedPrice = Number(utils.formatUnits(maximizedValue, SIGNED_PRICE_DECIMALS))
+    const minimizedPrice = Number(utils.formatUnits(minimizedValue, SIGNED_PRICE_DECIMALS))
+    const result = median([minimizedPrice, maximizedPrice])
+
+    return {
+      data: {
+        result,
+      },
+      statusCode: 200,
+      result,
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+  }
+
+  // Fetches the lwba price info from multiple source EAs, calculates the median for bids and asks per asset and fixes the price precision
+  private async fetchPrices(assets: string[], dataRequestedTimestamp: number) {
+    // priceData holds raw bid/ask values per asset from source EAs response
+    const priceData = {} as PriceData
+    assets.forEach((a) => {
+      priceData[a] = { bids: [], asks: [] }
+    })
```

Comment:
I don't think this is necessary, as we explicitly add the key on request response

```suggestion
```